crates/ika-core/src/dwallet_mpc/mpc_manager.rs

```diff
@@ -238,390 +211,99 @@ impl DWalletMPCManager {
-                    malicious_actors: vec![*authority],
-                }
-            });
-        Ok(output_verification_result)
-    }
-
-    fn handle_threshold_not_reached_report(
-        &mut self,
-        report: ThresholdNotReachedReport,
-        origin_authority: AuthorityName,
-    ) -> DwalletMPCResult<()> {
-        // Previously malicious actors are ignored.
-        if self
-            .malicious_handler
-            .get_malicious_actors_names()
-            .contains(&origin_authority)
-        {
-            return Ok(());
-        }
-        let committee = self.epoch_store()?.committee().clone();
-        let current_voters_for_report = self
-            .threshold_not_reached_reports
-            .entry(report.clone())
-            .or_insert(StakeAggregator::new(committee));
-        // We already have a quorum for this report.
-        if current_voters_for_report.has_quorum() {
-            // Do nothing, quorum has already been reached.
-            return Ok(());
-        }
-        if current_voters_for_report
-            .insert_generic(origin_authority, ())
-            .is_quorum_reached()
-        {
-            self.prepare_for_round_retry(report.session_identifier)?;
-        }
-        Ok(())
-    }
-
-    fn prepare_for_round_retry(
+    /// Handle the messages of a given consensus round.
+    pub fn handle_consensus_round_messages(
         &mut self,
-        session_identifier: SessionIdentifier,
-    ) -> DwalletMPCResult<()> {
-        let epoch_store = self.epoch_store()?;
-        if let Some(session) = self.mpc_sessions.get_mut(&session_identifier) {
-            session.attempts_count += 1;
-            // We got a `TWOPCMPCThresholdNotReached` error and a quorum agreement on it.
-            // So all parties that sent a regular MPC Message for the last executed
-            // round are malicious—as the round aborted with the error `TWOPCMPCThresholdNotReached`.
-            // All honest parties should report that there is a quorum for `ThresholdNotReached`.
-            // We must then remove these messages and mark the senders as malicious.
-            // Note that the current round was already incremented
-            // since we received the quorum for `ThresholdNotReached`
-            // on the previous round,
-            // but no messages were sent for the current round.
-            self.malicious_handler
-                .report_malicious_actors(&party_ids_to_authority_names(
-                    &session
-                        .serialized_full_messages
-                        .get(&(session.current_round - 1))
-                        .unwrap_or(&HashMap::new())
-                        .keys()
-                        .cloned()
-                        .collect::<Vec<PartyID>>(),
-                    &epoch_store,
-                )?);
+        consensus_round: u64,
+        messages: Vec<DWalletMPCMessage>,
+    ) {
+        for (_, session) in self.mpc_sessions.iter_mut() {
```

Comment:
The session can be created before by an event. You should only add rounds of messages or session with rounds